Source/WebGPU/WebGPU/Adapter.mm

```diff
@@ -85,33 +85,25 @@
 void Adapter::requestDevice(const WGPUDeviceDescriptor& descriptor, CompletionHandler<void(WGPURequestDeviceStatus, Ref<Device>&&, String&&)>&& callback)
 {
     if (descriptor.nextInChain) {
-        instance().scheduleWork([strongThis = Ref { *this }, callback = WTFMove(callback)]() mutable {
-            callback(WGPURequestDeviceStatus_Error, Device::createInvalid(strongThis), "Unknown descriptor type"_s);
-        });
+        callback(WGPURequestDeviceStatus_Error, Device::createInvalid(*this), "Unknown descriptor type"_s);
```

Comment:
I think this is backwards, and we should make requestAdapter() and requestDevice properly async (now that we can).

This change is probably okay if it's temporary, but let's at least file a bug about making those functions actually async.

(And ditto for below.)

Reply:
I've filed https://bugs.webkit.org/show_bug.cgi?id=251713 and will update the FIXME that exists below.

But in the case of failure, there is no reason the callback should be called async, we can return immediately and notify the client of the failure. This is inline with the web and native specifications.

I can understand the reasoning for making the callback asynchronous if the device creation takes a while, but then it should probably occurring off the main thread in the case of native webgpu

Reply:
> But in the case of failure, there is no reason the callback should be called async, we can return immediately and notify the client of the failure.

The [spec says](https://gpuweb.github.io/gpuweb/#dom-gpuadapter-requestdevice):
> 1. Let contentTimeline be the current [Content timeline](https://gpuweb.github.io/gpuweb/#content-timeline).
> 2. Let promise be [a new promise](https://webidl.spec.whatwg.org/#a-new-promise).
> 3. Let adapter be this.[[[adapter]]](https://gpuweb.github.io/gpuweb/#dom-gpuadapter-adapter-slot).
> 4. Issue the initialization steps to the [Device timeline](https://gpuweb.github.io/gpuweb/#device-timeline) of this.
> 5. Return promise.
Because validation happens on the device timeline, I think failure is still supposed to be reported asynchronously.

> I can understand the reasoning for making the callback asynchronous if the device creation takes a while, but then it should probably occurring off the main thread in the case of native webgpu

`MTLDevice`s are singletons, so creating one is not slow. `requestDevice()` does more than that, though: it sorts and filters devices, inspects them for family capabilities, and, creates a queue. It sounds worthy of a new bug to determine whether or not we get benefit by doing device creation on a secondary thread.

Reply:
Maybe I misunderstand but aren't the concepts of device and content pipelines specific to the web api and not the native api?

A native webgpu application could have everything on the main thread if it desired, in which case the two timelines are the same.

Reply:
I think this is the same discussion as https://github.com/WebKit/WebKit/pull/9584#discussion_r1096406506 - in the WebGPU API, it's impossible for the promise's `.then` callback to be called in the middle of `requestAdapter()`. That's the semantics I'm trying to preserve.